File: Majority Element/Majority_Element.cpp
// 对于这道题目，可以使用哈希表或者排序，但是还有一种特殊的算法，专门解决这类的问题
// Boyer-Moore投票算法

// Runtime: 20 ms, faster than 99.42% of C++ online submissions for Majority Element.
// Memory Usage: 11.1 MB, less than 100.00% of C++ online submissions for Majority Element.

class Solution 
{
public:
    int majorityElement(vector<int>& nums) 
    {
        int count = 0;
        int res;

        for (auto num : nums)
        {
            if (count == 0)
                res = num;
            
            if (res == num)
                count++;
            else
                count--;
        }
        return res;
    }
};

File: Majority Element/Majority_Element.py
# Runtime: 52 ms, faster than 73.65% of Python3 online submissions for Majority Element.
# Memory Usage: 13.5 MB, less than 100.00% of Python3 online submissions for Majority Element.

class Solution:
    def majorityElement(self, nums: 'List[int]') -> 'int':
        count = 0
        res = None

        for num in nums:
            if count is 0:
                res = num
            if res == num:
                count += 1
            else:
                count -= 1
        return res